airbyte-integrations/connectors/source-google-ads/source_google_ads/components.py

```diff
@@ -939,3 +896,212 @@ def _validate_query(self, query: str):
                 message=f"The provided query is invalid: {query}. Please refer to the Google Ads API documentation for the correct syntax: https://developers.google.com/google-ads/api/fields/v20/overview and test validate your query using the Google Ads Query Builder: https://developers.google.com/google-ads/api/fields/v20/query_validator",
             )
+
+
+@dataclass
+class StringParseState:
+    inside_string: bool = False
+    escape_next_character: bool = False
+    collected_string_chars: List[str] = field(default_factory=list)
+    last_parsed_key: Optional[str] = None
+
+
+@dataclass
+class TopLevelObjectState:
+    depth: int = 0
+
+
+@dataclass
+class ResultsArrayState:
+    inside_results_array: bool = False
+    array_nesting_depth: int = 0
+    expecting_results_array_start: bool = False
+
+
+@dataclass
+class RecordParseState:
+    inside_record: bool = False
+    record_text_buffer: List[str] = field(default_factory=list)
+    record_nesting_depth: int = 0
+
+
+@dataclass
+class GoogleAdsStreamingDecoder(Decoder):
+    """
+    JSON streaming decoder optimized for Google Ads API responses.
+
+    Uses a fast JSON parse when the full payload fits within MAX_DIRECT_DECODE_BYTES;
+    otherwise streams records incrementally from the `results` array.
+    Ensures truncated or structurally invalid JSON is detected and reported.
+    """
+
+    CHUNK_SIZE = 5 * 1024 * 1024  # 5 MB
+    # Fast-path threshold: if whole body ≤ 20 MB, decode with json.loads
+    MAX_DIRECT_DECODE_BYTES = 20 * 1024 * 1024  # 20 MB
+
+    def __post_init__(self):
+        self.parser = JsonParser()
+
+    def is_stream_response(self) -> bool:
+        return True
+
+    def decode(self, response: requests.Response) -> Generator[MutableMapping[str, Any], None, None]:
+        data, complete = self._buffer_up_to_limit(response)
+        if complete:
+            yield from self.parser.parse(io.BytesIO(data))
+            return
+
+        records_batch: List[Dict[str, Any]] = []
+        for record in self._parse_records_from_stream(data):
+            records_batch.append(record)
+            if len(records_batch) >= 100:
+                yield {"results": records_batch}
+                records_batch = []
+
+        if records_batch:
+            yield {"results": records_batch}
+
+    def _buffer_up_to_limit(self, response: requests.Response) -> Tuple[Union[bytes, Iterable[bytes]], bool]:
+        buf = bytearray()
+        response_stream = response.iter_content(chunk_size=self.CHUNK_SIZE)
+
+        while chunk := next(response_stream, None):
+            buf.extend(chunk)
+            if len(buf) > self.MAX_DIRECT_DECODE_BYTES:
```

Comment:
Because it is `>` and not `>=`, coes that mean that we will actually pull 25 Mb in memory instead of the threshold of 20 Mb described by `MAX_DIRECT_DECODE_BYTES`? I'm not actually opposed to pulling more in memory but I fear it might be confusing for the dev in case of a memory issue investigation